src/com/palmergames/bukkit/towny/utils/SpawnUtil.java

```diff
@@ -0,0 +1,395 @@
+			spawnPermission = String.format(spawnType.getTypeName() + " (%s)", nationSpawnPermission);
+			payee = nation;
+			break;
+		}
+
+		// Check if need/can pay.
+		try {
+			if ((!townyUniverse.getPermissionSource().has(player,
+					PermissionNodes.TOWNY_COMMAND_TOWNYADMIN_TOWN_SPAWN_FREECHARGE.getNode()))
+					&& (travelCost > 0 && TownySettings.isUsingEconomy()
+							&& (resident.getHoldingBalance() < travelCost)))
+				throw new TownyException(notAffordMSG);
+		} catch (EconomyException ignored) {
+		}
+
+		// Used later to make sure the chunk we teleport to is loaded.
+		Chunk chunk = spawnLoc.getChunk();
+
+		// Essentials tests.
+		boolean usingESS = plugin.isEssentials();
+
+		if (usingESS && !isTownyAdmin) {
+			try {
+				User essentialsUser = plugin.getEssentials().getUser(player);
+
+				// This jail check is specifically for essentials jails, not towny ones.
+				if (!essentialsUser.isJailed()) {
+
+					Teleport teleport = essentialsUser.getTeleport();
+					if (!chunk.isLoaded())
+						chunk.load();
+					// Cause an essentials exception if in cooldown.
+					teleport.cooldown(true);
+					teleport.teleport(spawnLoc, null, TeleportCause.COMMAND);
+				}
+			} catch (Exception e) {
+				TownyMessaging.sendErrorMsg(player, "Error: " + e.getMessage());
+				return;
+			}
+		}
+
+		// Actual taking of monies here.
+		if (!townyUniverse.getPermissionSource().has(player,
+				PermissionNodes.TOWNY_COMMAND_TOWNYADMIN_TOWN_SPAWN_FREECHARGE.getNode())) {
+			if (!TownySettings.isTownSpawnPaidToTown())
+				payee = TownyEconomyObject.SERVER_ACCOUNT;
+			// Show message if we are using an Economy and are charging for spawn travel.
+			try {
+				if (travelCost > 0 && TownySettings.isUsingEconomy()
+						&& resident.payTo(travelCost, payee, spawnPermission)) {
+					TownyMessaging.sendMsg(player, String.format(TownySettings.getLangString("msg_cost_spawn"),
+							TownyEconomyHandler.getFormattedBalance(travelCost)));
+				}
+			} catch (EconomyException ignored) {
+			}
+		}
+
+		// If an Admin or Essentials teleport isn't being used, use our own.
+		if (isTownyAdmin) {
+			if (player.getVehicle() != null)
+				player.getVehicle().eject();
+			if (!chunk.isLoaded())
+				chunk.load();
+			player.teleport(spawnLoc, TeleportCause.COMMAND);
+			return;
+		}
+
+		if (!usingESS) {
+			if (TownyTimerHandler.isTeleportWarmupRunning()) {
+				// Use teleport warmup
+				player.sendMessage(String.format(TownySettings.getLangString("msg_town_spawn_warmup"),
+						TownySettings.getTeleportWarmupTime()));
+				TownyAPI.getInstance().requestTeleport(player, spawnLoc);
+			} else {
+				// Don't use teleport warmup
+				if (player.getVehicle() != null)
+					player.getVehicle().eject();
+				if (!chunk.isLoaded())
+					chunk.load();
+				player.teleport(spawnLoc, TeleportCause.COMMAND);
+				if (TownySettings.getSpawnCooldownTime() > 0)
+					CooldownTimerTask.addCooldownTimer(resident.getName(), CooldownType.TELEPORT);
+			}
+		}

```

Comment:
This code can be cleaned up and make use of Paper servers with this replacement:

```java
		if (usingESS) {
			try {
				User essentialsUser = plugin.getEssentials().getUser(player);

				// This jail check is specifically for essentials jails, not towny ones.
				if (isTownyAdmin || !essentialsUser.isJailed()) {

					Teleport teleport = essentialsUser.getTeleport();
					
					// Cause an essentials exception if in cooldown.
					teleport.cooldown(true);
					teleport.teleport(spawnLoc, null, TeleportCause.COMMAND);
				}
			} catch (Exception e) {
				TownyMessaging.sendErrorMsg(player, "Error: " + e.getMessage());
				return;
			}
		} else if (!isTownyAdmin && TownyTimerHandler.isTeleportWarmupRunning()) {
			// Use teleport warmup
			player.sendMessage(String.format(TownySettings.getLangString("msg_town_spawn_warmup"), TownySettings.getTeleportWarmupTime()));
			TownyAPI.getInstance().requestTeleport(player, spawnLoc);
			return;
		} else {
			// Don't use teleport warmup
			if (player.getVehicle() != null)
				player.getVehicle().eject();

			if (!chunk.isLoaded())
				chunk.load();
			player.teleport(spawnLoc, TeleportCause.COMMAND);
		}

		if (!isTownyAdmin && TownySettings.getSpawnCooldownTime() > 0)
			CooldownTimerTask.addCooldownTimer(resident.getName(), CooldownType.TELEPORT);
```